content/blog/magrittr-2-0/index.Rmd

```diff
@@ -0,0 +1,195 @@
+#>  2. │ └─global::bar()
+#>  3. │   └─1 %>% identity() %>% baz()
+#>  4. └─global::baz(.)
+```
+
+Note however that one consequence of having a lazy pipe is that the whole pipeline is pushed on the stack before errors have a chance to be thrown.
+
+```{r, eval = FALSE}
+f1 <- function(x) x
+f2 <- function(x) x
+f3 <- function(x) x
+f4 <- function(x) x
+
+stop("oh no") %>% f1() %>% f2() %>% f3() %>% f4()
+#> Error in f1(.) : oh no
+
+rlang::last_trace()
+#> <error/rlang_error>
+#> oh no
+#> Backtrace:
+#>     █
+#>  1. ├─stop("oh no") %>% f1() %>% f2() %>% f3() %>% f4()
+#>  2. ├─global::f4(.)
+#>  3. ├─global::f3(.)
+#>  4. ├─global::f2(.)
+#>  5. └─global::f1(.)
+```
+
+The last function of the pipeline is `f4()`, so that's the first one to be run. It evaluates its argument which is provided by `f3()`, so that's the second function pushed on the stack. And so on until `f1()` needs the result of `stop("oh no")` which causes an error.
+
+
+## Speed
+
+The pipe is now written in C. This greatly improves the performance. Here is a benchmark for the old R implementation:
+
+```{r, eval = FALSE}
+bench::mark(
+  `0` = f1(NULL),
+  `1` = NULL %>% f1(),
+  `2` = NULL %>% f1() %>% f2(),
+  `3` = NULL %>% f1() %>% f2() %>% f3(),
+  `4` = NULL %>% f1() %>% f2() %>% f3() %>% f4(),
+)
+#> + + + + + + # A tibble: 5 x 13
+#>   expression     min  median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result
+#>   <bch:expr> <bch:t> <bch:t>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list>
+#> 1 0            258ns   319ns  2808344.        0B      0   10000     0     3.56ms <NULL>
+#> 2 1           59.4µs  68.9µs    13648.      280B     59.1  6004    26   439.91ms <NULL>
+#> 3 2           82.6µs 101.6µs     9252.      280B     42.8  3894    18   420.87ms <NULL>
+#> 4 3          106.4µs 124.7µs     7693.      280B     18.8  3690     9   479.64ms <NULL>
+#> 5 4          130.9µs 156.1µs     6173.      280B     18.8  2956     9   478.84ms <NULL>
+#> # … with 3 more variables: memory <list>, time <list>, gc <list>
+```
+
+The new C implementation is less costly per pipe expression:
+
+```{r, eval = FALSE}
+bench::mark(
+  `0` = f1(NULL),
+  `1` = NULL %>% f1(),
+  `2` = NULL %>% f1() %>% f2(),
+  `3` = NULL %>% f1() %>% f2() %>% f3(),
+  `4` = NULL %>% f1() %>% f2() %>% f3() %>% f4(),
+)
+#>   expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result
+#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list>
+#> 1 0             270ns    383ns  2240689.        0B    224.   9999     1     4.46ms <NULL>
+#> 2 1            4.47µs   5.95µs   159655.        0B     79.9  9995     5     62.6ms <NULL>
+#> 3 2            5.97µs    8.8µs   109534.        0B     32.9  9997     3    91.27ms <NULL>
+#> 4 3            8.83µs  10.63µs    89902.        0B     27.0  9997     3    111.2ms <NULL>
+#> 5 4           10.99µs  13.18µs    72330.        0B     36.2  9995     5   138.19ms <NULL>
+```
+
```

Comment:
We don't generally except this to have much impact on data analysis code, but it might yield meaningful speed ups if you had a pipe inside a tight loop.